### ISSUE TITLE
Difference between the SharePoint API Management and the Azure AD App Registrations blade

### ISSUE THREAD

**Issue by colinrippeyfinarne** [SharePoint/sp-dev-docs#2516]
What is the difference between granting permissions on the SharePoint Admin screen versus granting permissions directly in the Azure portal via the Azure AD App Registrations blade?

I am struggling to understand why SharePoint has its own API approval screen. Does approving a permission from the SharePoint admin screen actually add the permission to the underlying App Registration in the Azure Portal or are these two screens completely separate from each other in terms of the information they present.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 47b09a1d-5520-82e2-64f5-7aaae817720a
* Version Independent ID: 0e3dcdc4-3a49-b4c7-f525-1c618cf5e21e
* Content: [Connect to Azure AD-secured APIs in SharePoint Framework solutions](https://docs.microsoft.com/en-us/sharepoint/dev/spfx/use-aadhttpclient#manage-permission-requests)
* Content Source: [docs/spfx/use-aadhttpclient.md](https://github.com/SharePoint/sp-dev-docs/blob/master/docs/spfx/use-aadhttpclient.md)
* Product: **sharepoint**
* GitHub Login: @spdevdocs
* Microsoft Alias: **spdevdocs**

**Comment by waldekmastykarz** [comment]
When you grant permissions directly in Azure AD, you need to register a new application and in SharePoint implement OAuth implicit flow for that application to obtain an access token. The challenge you have, is that you cannot know upfront on which pages your solution will run, which you must know, because the full URL is a part of the OAuth implicit flow contract you register in Azure AD.

When granting permissions through SharePoint Admin screen, you leverage the Azure AD principal provisioned by Microsoft for which Microsoft implemented OAuth implicit flow in a way that doesn't require you to register all the pages where your components are used. Each permission you grant, is added to the existing principal.

So when comparing both approaches, granting permissions via SharePoint Admin simplifies deployment and management of permissions and gives SharePoint admins (not Azure AD admins) insights into what permissions are granted to scripts in SharePoint.

**Comment by vman** [comment]
To add to what @waldekmastykarz said, having an API management page living in SharePoint also helps in cases where organisations don't have/want Azure rolled out to the business yet. In these cases, the deployment of SPFx solutions does not get blocked as the permissions can be managed from SharePoint itself.

**Comment by colinrippeyfinarne** [comment]
Thanks for the responses.

I guess I am trying to correlate what actually happens at the Azure end when a SharePoint Admin clicks the Approve API Permission request.

I am finding that the permissions granted via the SharePoint API management screen show up in the Azure AD App Reg blade for the spfx App Reg, but not for every single permission grant (some permission grants to custom Azure Function App Regs don't seem to show up in Azure AD, they appear as approved in the SharePoint API management screen and I can call them via a spfx webpart using AadHttpClient, but they don't show up in the Azure AD end and so I am still confused as to how these permission grants from SharePoint API Permissions are actually resolved at the Azure AD end when the authentication is done?).

I am also finding that if I go to Azure AD directly, App Registrations, and then manage the permissions on the Spfx App Reg by say adding a permission to call some graph apis or even a custom Azure Function (secured with Azure AD) that these permissions then show up as Approved in the SharePoint Admin API management screen (and also they show up if you run the Get-SPOTenantServicePrincipalPermissionGrants commandlet). 

So it seems I can bypass adding the webApiPermissionRequests to my package-solution.json and manage the permissions separately, but I don't know if I am missing something here...does adding the permissions at the Azure AD end miss out some step that adding at the SharePoint API management screen does?

My experience with testing out spfx packages is such that my SharePoint API management list is now cluttered with the same list of permissions from the same .sppkg file as I go through a loop of "dev,test,deploy,test" etc., I have deployed the same package multiple times and in doing so the list of API Permission requests is ever growing and I wouldn't really know what ones are "relevant" or can be removed?

Given that every Office 365 tenant has an Azure AD directory that is easily accessible from the Office 365 Admin screen I wonder if it is simpler for (at least) developers to consider managing their permissions via Azure AD?

I also want to flesh out how the granting of custom apis in the webApiPermissionRequests seems to work but no actual permissions are present on the spfx app reg, how does this work in practise? There seems to be two lists of permissions...the Get-SPOTenantServicePrincipalPermissionGrants seems to show me all the permissions regardless of whether they've been added as part of a sppkg package or have been manually added to the Azure AD spfx App Reg.

How does the spfx app reg "know" it has permission to call the Azure Function API? Does SharePoint maintain a different list? Does the spfx App Reg look at its permissions in Azure AD and also another list of permission grants maintained by SharePoint?

**Comment by waldekmastykarz** [comment]
@colinrippeyfinarne 

> I am finding that the permissions granted via the SharePoint API management screen show up in the Azure AD App Reg blade for the spfx App Reg, but not for every single permission grant (some permission grants to custom Azure Function App Regs don't seem to show up in Azure AD, they appear as approved in the SharePoint API management screen and I can call them via a spfx webpart using AadHttpClient, but they don't show up in the Azure AD end and so I am still confused as to how these permission grants from SharePoint API Permissions are actually resolved at the Azure AD end when the authentication is done?).

This is odd. Are you sure that you're looking at the right principal in AAD?

> So it seems I can bypass adding the webApiPermissionRequests to my package-solution.json and manage the permissions separately, but I don't know if I am missing something here...does adding the permissions at the Azure AD end miss out some step that adding at the SharePoint API management screen does?

You're not missing anything and this is exactly how it works. When you deploy a package with permission requests to app catalog, an entry is added to the Web API permissions list for each request as stated in the package. When you approve a request, it gets assigned to the principal in AAD. In fact, in an enterprise organization you would do exactly that: enable a predefined set of permissions upfront and ignore permissions requests issued by packages. That way you would have more control about what permissions your organization is ok to open to solutions.

> Given that every Office 365 tenant has an Azure AD directory that is easily accessible from the Office 365 Admin screen I wonder if it is simpler for (at least) developers to consider managing their permissions via Azure AD?

Some organizations have separate teams being responsible for SharePoint and Azure so this might not work for everybody.

> How does the spfx app reg "know" it has permission to call the Azure Function API? Does SharePoint maintain a different list? Does the spfx App Reg look at its permissions in Azure AD and also another list of permission grants maintained by SharePoint?

There are two lists of permissions:
- webApiPermissionRequests: this is a list of permissions requested to be granted. This list exists only in SharePoint and is not visible or stored anywhere except the SharePoint list
- granted API permissions: these permissions are the actual permissions assigned to the specific principal in the AAD and they are shown in the API management under approved permissions or using PowerShell.

> How does the spfx app reg "know" it has permission to call the Azure Function API?

When requesting an access token to an API, the request is done on behalf of the specific principal that has all the permissions associated with it. If the principal has no permissions granted for the particular resource (API), you will get an error when trying to request an access token. All this is handled by AAD and OAuth and there is nothing specific to SharePoint in it.

Hope this helps

**Comment by vman** [comment]
Hi @colinrippeyfinarne When you use the SharePoint API Management page to grant permissions, they are added to the Service Principal and not the App Registration. I believe you are not able to see the permissions granted because you are looking in Azure AD -> App Registrations.

If you look at the Service Principal by going to Azure Active Directory -> Enterprise Applications -> All Applications -> SharePoint Online Client Extensibility Web Application Principal -> Permissions

you should be able to see the permissions there.

Perhaps @patmill or @lucaband can help us out here on why the permissions are not added to the App Registration as well?

**Comment by colinrippeyfinarne** [comment]
> This is odd. Are you sure that you're looking at the right principal in AAD?

I've tried to completely re-create this exact scenario on a totally separate tenant...and yep I see the same thing again only this time I also don't see the Microsoft Graph permissions on the Spfx App Reg.

I've got a simple Azure Function with this code:

`    using System.Collections.Generic;
     using System.Net;
     using System.Net.Http;
     using System.Net.Http.Formatting;
    using System.Security.Claims;
    public static async Task<HttpResponseMessage> Run(HttpRequestMessage req, TraceWriter log)
    {
        log.Info("C# HTTP trigger function processed a request.");
        log.Info("C# HTTP trigger function processed a request.");
        var claimsList = new Dictionary<string, string>();
        claimsList.Add("App Reg Name", "Spfx WebPart via webApiPermissionRequests");

        //Current user claims
        foreach (Claim claim in ClaimsPrincipal.Current.Claims)
        {
            string claimType = claim.Type;
            string claimValue = claim.Value;
            result = result + "-" + "Type " + claimType + " Value " + claimValue;
            claimsList.Add(claimType, claimValue);
        }

        var claimsFound = ClaimsPrincipal.Current.Claims;
        var didWeFind = claimsFound.GetType().Name;

        return req.CreateResponse(HttpStatusCode.OK, claimsList, JsonMediaTypeFormatter.DefaultMediaType);
    }
`
I then secured this function using Azure AD with an App Reg (I used the Express config, used the following display name)

ApplicationId: 6b9d6cfb-584e-4ea1-a5f3-e7c13f820464
DisplayName: Spfx WebPart Auto DisplayName

I configured the function to have a CORS entry for the SharePoint domain.

I deliberately did NOT go into the Azure AD App Reg blade for the App Reg and grant admin consent to the permissions on this new App Reg...left that alone for now (as I was wondering if this would cause the authentication to break and wanted to see if I had to turn on one bit at a time).

I created a simple spfx webpart with the following entries in the package-solution.json file (I modified the online sample from :

    "webApiPermissionRequests": [
      {
        "resource": "Microsoft Graph",
        "scope": "User.ReadBasic.All"
      },
      {
        "resource": "Spfx WebPart Auto DisplayName",
        "scope": "user_impersonation"
      }
    ]

The spfx webpart tsx file has the following code (again a simple modification from the above sample):

    `  
    private _searchWithAad(): void {

    // Log the current operation
    console.log("Using _searchWithAad() method");

    // Using Graph here, but any 1st or 3rd party REST API that requires Azure AD auth can be used here.
    this.props.context.aadHttpClientFactory
      .getClient('6b9d6cfb-584e-4ea1-a5f3-e7c13f820464')
      .then((client: AadHttpClient) => {
        return client
          .get('https://myfunctionname.azurewebsites.net/api/CallableFromSpfxAutoDisplayName',
          AadHttpClient.configurations.v1
          );
      })
      .then(response => {
        return response.json();
      })
      .then(json => {
        console.log(json);        
      });

    this.props.context.aadHttpClientFactory
      .getClient('https://graph.microsoft.com')
      .then((client: AadHttpClient) => {
        // Search for the users with givenName, surname, or displayName equal to the searchFor value
        return client
          .get(
            `https://graph.microsoft.com/v1.0/users?$select=displayName,mail,userPrincipalName&$filter=(givenName%20eq%20'${escape(this.state.searchFor)}')%20or%20(surname%20eq%20'${escape(this.state.searchFor)}')%20or%20(displayName%20eq%20'${escape(this.state.searchFor)}')`,
            AadHttpClient.configurations.v1
          );
      })
      .then(response => {
        return response.json();
      })
      .then(json => {

        // Prepare the output array
        var users: Array<IUserItem> = new Array<IUserItem>();

        // Log the result in the console for testing purposes
        console.log(json);

        // Map the JSON response to the output array
        json.value.map((item: any) => {
          users.push( {
            displayName: item.displayName,
            mail: item.mail,
            userPrincipalName: item.userPrincipalName,
          });
        });

        // Update the component state accordingly to the result
        this.setState(
          {
            users: users,
          }
        );
      })
      .catch(error => {
        console.error(error);
      });
    }
    `
I've packaged the solution:

gulp build
gulp bundle
gulp package-solution

Uploaded to the tenant app catalog, deployed, approved the API Permissions.

If I run the Get-SPOTenantServicePrincipalPermissionGrants I see:

ClientId    : 829c1b76-c61e-478a-bd69-a0ef8ea6afd7
ConsentType : AllPrincipals
ObjectId    : dhucgh7Gike9aaDvjqav1_pMvFTqzbNPkzPSC9ZwQfY
Resource    : Microsoft Graph
ResourceId  : 54bc4cfa-cdea-4fb3-9333-d20bd67041f6
Scope       : User.ReadBasic.All

ClientId    : 829c1b76-c61e-478a-bd69-a0ef8ea6afd7
ConsentType : AllPrincipals
ObjectId    : dhucgh7Gike9aaDvjqav14qkCKB1o_hDoiwOFqPL4O0
Resource    : Spfx WebPart Auto DisplayName
ResourceId  : a008a48a-a375-43f8-a22c-0e16a3cbe0ed
Scope       : user_impersonation

So...it would appear that the API Requests are all good to go.

I run 

gulp serve --nobrowser

(is this the issue...am I confusing everything by running the local webpart???)

I then go to a modern team site and browse to the workbench.aspx and add the webpart.

I use the simple search button to cause the webpart code to run, I get successful results (the sample webpart presents the details of a user it finds)

From the console for the output of the function:

FunctionAppReg: "Spfx WebPart Auto DisplayName"
aio: "42BgYPg6VbCLp/Z3zCnm5hPioeuzLiuqx8p/5ez6uXe1x0He2tUA"
appid: "5d437051-7e4b-418e-b940-2f3385777a67"
appidacr: "0"
aud: "6b9d6cfb-584e-4ea1-a5f3-e7c13f820464"
e_exp: "262800"
exp: "1536848981"
http://schemas.microsoft.com/claims/authnclassreference: "1"
http://schemas.microsoft.com/claims/authnmethodsreferences: "pwd"
http://schemas.microsoft.com/identity/claims/objectidentifier: "ff8aac10-b65e-4991-bf47-daf92ef4f555"
http://schemas.microsoft.com/identity/claims/scope: "user_impersonation"
http://schemas.microsoft.com/identity/claims/tenantid: "e10a310a-af25-474d-b5b0-1acab1e4483c"
http://schemas.xmlsoap.org/ws/2005/05/identity/claims/givenname: "Colin"
http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name: "colin.rippey@mytenant.onmicrosoft.com"
http://schemas.xmlsoap.org/ws/2005/05/identity/claims/nameidentifier: "D7v7SqQzzmnFQpFLnQYOxXW47x5e4QPYThoiU2rc150"
http://schemas.xmlsoap.org/ws/2005/05/identity/claims/surname: "Rippey"
http://schemas.xmlsoap.org/ws/2005/05/identity/claims/upn: "colin.rippey@mytenant.onmicrosoft.com"
iat: "1536845081"
ipaddr: "108.171.128.174"
iss: "https://sts.windows.net/e10a310a-af25-474d-b5b0-1acab1e4483c/"
name: "Colin Rippey"
nbf: "1536845081"
uti: "DcTndB_f6Ui8p5DNXwyFAA"
ver: "1.0"

So whatever I've done seems to have granted permissions on the Azure AD portal to allow the client code to execute against the function.

As I understand it the way this all works is the Spfx App Reg in my tenant: 

Application Id: 5d437051-7e4b-418e-b940-2f3385777a67
Display Name: SharePoint Online Client Extensibility Web Application Principal

Is what is being used to make the authentication flow through to the function.

My main question here...is should I now see the permissions on the Azure AD App Reg:

Application Id: 5d437051-7e4b-418e-b940-2f3385777a67
Display Name: SharePoint Online Client Extensibility Web Application Principal

If I browse to this App Reg in the Azure Portal I don't see any permissions at all!

If I go to Graph Explorer and issue the following:

https://graph.microsoft.com/beta/serviceprincipals?$filter=startswith(displayName,'SharePoint Online Client Extensibility Web Application Principal')

I see:

    `
    {
    "@odata.context": "https://graph.microsoft.com/beta/$metadata#servicePrincipals",
    "value": [
        {
            "id": "829c1b76-c61e-478a-bd69-a0ef8ea6afd7",
            "deletedDateTime": null,
            "accountEnabled": true,
            "appDisplayName": "SharePoint Online Client Extensibility Web Application Principal",
            "appId": "5d437051-7e4b-418e-b940-2f3385777a67",
            "appOwnerOrganizationId": "e10a310a-af25-474d-b5b0-1acab1e4483c",
            "appRoleAssignmentRequired": false,
            "displayName": "SharePoint Online Client Extensibility Web Application Principal",
            "errorUrl": null,
            "homepage": null,
            "logoutUrl": null,
            "publishedPermissionScopes": [
                {
                    "adminConsentDescription": "Allow the application to access SharePoint Online Client Extensibility Web Application Principal on behalf of the signed-in user.",
                    "adminConsentDisplayName": "Access SharePoint Online Client Extensibility Web Application Principal",
                    "id": "272813c6-9a34-46e5-aed4-0b4bc74cdff3",
                    "isEnabled": true,
                    "type": "User",
                    "userConsentDescription": "Allow the application to access SharePoint Online Client Extensibility Web Application Principal on your behalf.",
                    "userConsentDisplayName": "Access SharePoint Online Client Extensibility Web Application Principal",
                    "value": "user_impersonation"
                }
            ],
            "preferredTokenSigningKeyThumbprint": null,
            "publisherName": "sqanowpersonaldev",
            "replyUrls": [
                "https://sqanowpersonaldev.sharepoint.com/",
                "https://sqanowpersonaldev.sharepoint.com/_forms/singlesignon.aspx",
                "https://sqanowpersonaldev.sharepoint.com/_forms/singlesignon.aspx?redirect",
                "https://sqanowpersonaldev.sharepoint.com/_forms/spfxsinglesignon.aspx",
                "https://sqanowpersonaldev.sharepoint.com/_forms/spfxsinglesignon.aspx?redirect"
            ],
            "samlMetadataUrl": null,
            "servicePrincipalNames": [
                "5d437051-7e4b-418e-b940-2f3385777a67"
            ],
            "signInAudience": "AzureADMyOrg",
            "tags": [],
            "addIns": [],
            "appRoles": [],
            "keyCredentials": [],
            "passwordCredentials": []
        }
    ]
    }
    `
So I am totally confused as to where I am looking and what I think I am supposed to see!

I thought "what if there is some issue with me running the package locally via the online workbench" so I did:

gulp bundle --ship
gulp package-solution --ship

Redeployed the package to the tenant app catalog, approved the deployment and then sorted out the API requests.

I then added the webpart to a page...tested...got the same successful results.

This is too confusing for me! I seem to have managed to deploy code to SharePoint that does not have the permissions inside Azure AD but the code runs successfully. 

SharePoint tells me the permissions are okay...Azure AD says otherwise.

**Comment by lucaband** [comment]
So.. it's really what Waldek said.. but you might be in a complex situation because you had that tenant before 1.6 and.. during that "dev preview" time, we had a different application principal.

in a standard fresh "after GA" situation that's what you will have:
* in AAD you will have an App Principal called  "SharePoint Online Client Extensibility Web Application Principal "
* you will have only one app principal of that type
* the App ID of that principal will be different for each tenant as that is not a 1st party MT App Principal but a per tenant app principal

we provision that principal the first time you land on the web APIs management page *or* when you use PowerShell to approve permissions that came from the package. 
The web APIs Management page is, as @waldekmastykarz said, just an helper page to allow SharePoint administrators to interact with AAD without requiring them to learn a new tool / dashboard. 
we do not prevent the admins to bypass that page and go directly to AAD. as a matter of fact, we do use AADGraph behind the curtains to stamp those scopes and provide admin consent to the app principal mentioned before.

Hope that makes sense!

**Comment by waldekmastykarz** [comment]
I'd say that the intention is that the API management page and the SharePoint Online PowerShell cmdlets should give you a consistent image of the granted permissions. Checking the permissions via the AAD portal doesn't work because the permissions are granted to the service principal not the actual app, which is why the permissions that are granted aren't surfaced in the UI. You can check that the specify permissions are granted on the AAD object representing the service principal using the Office 365 CLI which gets this information from AAD: https://blog.mastykarz.nl/grant-api-permissions-office-365-without-sharepoint-framework-packages/

**Comment by colinrippeyfinarne** [comment]
@waldekmastykarz Thanks, I'll go over your article and see what details are uncovered for my scenarios.

I'm "guessing" it's easy to find out what the Office 365 CLI is actually calling back into? Is it a clever wrapper around the graph?

**Comment by waldekmastykarz** [comment]
@colinrippeyfinarne yes, you can have a look in the Office 365 CLI code. It's calling the Azure AD Graph (not MS Graph) to get the information about the service principals and their OAuth2 grants. All REST calls so should be easy to follow.

**Comment by lucaband** [comment]
"And yet the first thing we need to do for the new spfx service principal is to grant it the permission Windows Azure Active Directry->User.Read and the logical place to grant this permission seems to be the Azure AD App Reg blade."

that was a temporary workaround ... it should no longer be required starting next week.

**Comment by colinrippeyfinarne** [comment]
Thank you for all the comments on this.

I've tried to digest all of this over the weekend and I now have the following thoughts.

The SharePoint Admin API management screen is correctly reading and writing the permissions against the spfx service principal. These permissions are what allows the OAuth Implicit flow to work.

The Azure AD App Reg blade is surfacing permissions against the Application object, but to me it should also acknowledge the permissions on the underlying/associated service principal, and this is an issue I feel I need to now raise with the Azure AD team as when I set permissions via this GUI it overwrites the permissions that the SharePoint Admin API management screen is reading from.

The Microsoft Graph API  https://graph.microsoft.com/beta/applications should also acknowledge the permissions on the underlying/associated service principal.

I will try to raise this with the Azure AD team.
